homeassistant/helpers/category_registry.py

```diff
@@ -0,0 +1,213 @@
+"""Provide a way to categorize things within a defined scope."""
+from __future__ import annotations
+
+from collections.abc import Iterable
+import dataclasses
+from dataclasses import dataclass, field
+from typing import Literal, TypedDict, cast
+
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.util.ulid import ulid_now
+
+from .typing import UNDEFINED, EventType, UndefinedType
+
+DATA_REGISTRY = "category_registry"
+EVENT_CATEGORY_REGISTRY_UPDATED = "category_registry_updated"
+STORAGE_KEY = "core.category_registry"
+STORAGE_VERSION_MAJOR = 1
+SAVE_DELAY = 10
+
+
+class EventCategoryRegistryUpdatedData(TypedDict):
+    """Event data for when the category registry is updated."""
+
+    action: Literal["create", "remove", "update"]
+    scope: str
+    category_id: str
+
+
+EventCategoryRegistryUpdated = EventType[EventCategoryRegistryUpdatedData]
+
+
+@dataclass(slots=True, kw_only=True, frozen=True)
+class CategoryEntry:
+    """Category registry entry."""
+
+    category_id: str = field(default_factory=ulid_now)
+    icon: str | None = None
+    name: str
+
+
+class CategoryRegistry:
```

Comment:
There was a base class for registries added a while back that could be used here

Reply:
"a while back" 

![CleanShot 2024-03-14 at 20 51 34@2x](https://github.com/home-assistant/core/assets/195327/ed65b5b3-72d6-42a6-8755-40444990bdd5)

🤣

Reply:
done in [ef197e3](https://github.com/home-assistant/core/pull/110897/commits/ef197e3338ae4e5015268486bd6e6cf771d6aee2)

Reply:
It felt longer.  😉